File: c++/driver2.cpp
#include "matrix.h"
#include "losses.h"
#include "layers.h"
#include "metrics.h"
#include "optimizers.h"
#include "data/data1000.h"

using namespace matrix;

int main(int argc, char *argv[]) {
	int N = 100;
	if (argc > 1) {
		N = std::stoi(argv[1]);
		if (N > 1000) {
			N = 1000;
		}
	}

	std::cout << "N: " << N << std::endl;
	int train_size = N * 0.8;
	int test_size = N - train_size;
    Matrix<double> x_train(train_size, 2);
    x_train.set_matrix(std::vector<std::vector<double>>(x_train_raw.begin(), x_train_raw.end() - 800 + train_size));
    Matrix<double> y_train(train_size, 2);
    y_train.set_matrix(std::vector<std::vector<double>>(y_train_raw.begin(), y_train_raw.end() - 800 + train_size));

    Matrix<double> x_test(test_size, 2);
    x_test.set_matrix(std::vector<std::vector<double>>(x_test_raw.begin(), x_test_raw.end() - 200 + test_size));
    Matrix<double> y_test(test_size, 2);
    y_test.set_matrix(std::vector<std::vector<double>>(y_test_raw.begin(), y_test_raw.end() - 200 + test_size));

    Dense layer1(2, 32);
    ReLU layer2;
    Dense layer3(32, 2);
    SoftmaxCrossEntropy layer4;
    optimizer::SGD sgd(1.0, 0.001);

    for (int i = 0; i < 1801; i++) {

	Matrix<double> out1 = layer1.forward(x_train);
	Matrix<double> out2 = layer2.forward(out1);
	Matrix<double> out3 = layer3.forward(out2);
	Matrix<double> out4 = layer4.forward(out3, y_train);
	double loss = layer4.get_loss();
	double acc = metric::accuracy(y_train, out4);

	Matrix<double> back4 = layer4.backward(out4, y_train);
	Matrix<double> back3 = layer3.backward(out2, back4);
	Matrix<double> back2 = layer2.backward(out1, back3);
	Matrix<double> back1 = layer1.backward(x_train, back2);
	
	
	sgd.pre_update();
	sgd.update(&layer3);
	sgd.update(&layer1);
	sgd.post_update();
	if (i % 100 == 0) {
	    Matrix<double> outtest1 = layer1.forward(x_test);
	    Matrix<double> outtest2 = layer2.forward(outtest1);
	    Matrix<double> outtest3 = layer3.forward(outtest2);
	    Matrix<double> outtest4 = layer4.forward(outtest3, y_test);
	    double losstest = layer4.get_loss();
	    double acctest = metric::accuracy(y_test, outtest4);

	    std::cout << "epoch: " << i;
	    std::cout << ", acc: " << acc;
	    std::cout << ", loss: " << loss;
	    std::cout << ", acc_test: " << acctest;
	    std::cout << ", loss_test: " << losstest;
	    std::cout << ", lr: " << std::fixed << sgd.get_lr() << std::endl;
	}

    }


    return 1;
}

File: python/driver2.py
from Layers import *
from Losses import *
from Metrics import *
from Optimizers import *
from data.data import *

'''
W = [[-0.05078057],
 [ 0.64329567],
 [ 1.20661244],
 [ 0.67322229]]

[[-0.05078057, -0.45289629],
 [ 0.64329567,  0.69494955],
 [ 1.20661244,  0.107874  ],
 [ 0.67322229,  0.65965736]]
'''
# X = [
#     [1.0, 2.0, 3.0],
#     [-4.0, -5.0, -6.0],
#     [7.0, 8.0, 9.0],
# ]

# y_true = [
#     [0, 1],
#     [1, 0],
#     [0, 1]
#     ]

x_train = np.array(x_train_raw_1000)
y_train = np.array(y_train_raw_1000)


x_test = np.array(x_test_raw_1000)
y_test = np.array(y_test_raw_1000)

layer1 = Dense(2, 16)
layer2 = ReLU()
layer3 = Dense(16, 2)
layer4 = SoftmaxCrossEntropy()
sgd = SGD(learning_rate = 1, decay=0.001)

for epoch in range(1001):
    out = layer1.forward(x_train)
    out2 = layer2.forward(out)
    out3 = layer3.forward(out2)
    out4 = layer4.forward(out3, y_train)
    loss = layer4.loss
    acc = Accuracy().calculate(y_train, out4)

    back4 = layer4.backward(out4, y_train)
    back3 = layer3.backward(back4)
    back3 = layer2.backward(back3)
    back1 = layer1.backward(back3)

    sgd.pre_update()
    sgd.update(layer3)
    sgd.update(layer1)
    sgd.post_update()

    if not epoch % 100:
        
        out = layer1.forward(x_test)
        out2 = layer2.forward(out)
        out3 = layer3.forward(out2)
        out4 = layer4.forward(out3, y_test)
        lossTest = layer4.loss
        accTest = Accuracy().calculate(y_test, out4)
        
        print(f'epoch: {epoch}, ' +
        f'acc: {acc:.3f}, ' +
        f'loss: {loss:.3f}, ' +
        f'acc_test: {accTest:.3f}, ' +
        f'loss_test: {lossTest:.3f}, ' +
        f'lr: {sgd.current_learning_rate:.3f}')
git commit -m "Python: Added SGD optimizer, added hardcoded data file, updated layers, added new driver for new data."